L2 announcement

dylandreimerink opened the pull request:
This PR adds the L2 announcement feature. This feature makes service VIPs (specifically ExternalIPs and Loadbalancer IPs) available/routable on a local area network by responding to ARP queries. For Cilium setups in home, office, or campus networks the only option to expose a service externally would have been via node ports. Node ports are limited in the ports they can use and do not provide out of the box high availability, making clients responsible for picking another host of the first option goes down. With L2 announcements, the same IP will fail over to a new node if the old node dies. It also offers more flexibility with regards to the choice of IPs and ports.

Due to the nature of ARP, only 1 node can respond to ARP queries with its MAC address. If multiple nodes are configured to announce the same IP, leader election will take place to decided which node should respond.

This is a block diagram of the implementation:
![l2-aware-lb-eBPF responder-v3 drawio](https://github.com/cilium/cilium/assets/1799415/594d603b-2503-4cbd-87b5-f0826c88559e)

Orange nodes are newly added, blue is existing and unmodified, green are existing and modified, gray/dotted are planned for future iterations.

To control the feature a new CiliumL2AnnouncementPolicy CRD is added, its resource, services, local node, and available network devices feed into the "L2 announcer" cell in the agent. This cell contains the logic that distills all k8s resources down into a desired state for the local datapath.

Communication happens via a statedb table which gets read by the "L2 responder" cell. Its job is to reconcile the desired state with the actual map state. And as soon as the GARP sender component is available it will trigger gratuitous ARP messages to be sent when first adding a given entry to the map.

A "L2 responder map" cell provides a wrapper around the actual BPF map interface. This map is loaded from the BPF map definition and used by the L2 responder code in BPF. The L2 responder code is called from the "from-netdev" program, so when L2 announcements are enabled, netdev programs are loaded and attached.

This feature is currently IPv4 only, but the design allows for the future addition of IPv6 support.

```release-note
Added L2 announcement feature
```

dylandreimerink committed:
config: Added options and helm values for L2 announcement feature

This commit adds flags, options and helm values to enable the new
L2 announcement feature and tune some of its parameters. These options
will be used by code added in followup commits in the patch set.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

dylandreimerink committed:
datapath,maps: Added agent liveness configmap value and cell

The L2 responder code needs to know if the agent stopped maintaining
map values so it can yield when the current node is no longer the
lease holder.

This commit adds a new value to the configmap and a cell that updates
the that value periodically with the current monotonic time. This is
a generic mechanism which can also be used by other feature in the
future, but for now is primarily intended for the above described
purpose, which will be implemented in a followup commit in this patch
set.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

dylandreimerink committed:
bpf,tests: Added ARP layer to pktgen

This commit adds logic for the ARP layer to the pktgen which will be
utilized by future tests for the L2 responder logic.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

dylandreimerink committed:
k8s: Add CiliumL2AnnouncementPolicy CRD

This commit adds the L2 Announcement Policy CRD for the new
L2 announcement feature which will be utilized in followup commits.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

dylandreimerink committed:
statedb,datapath: Make pkg/datapath/tables and add L2 announce table

This commit makes a new package where table definitions will live that
serve as communication layer between datapath and agent components.

The first statedb table to be added is the L2 announce table. Each entry
contains a network device and IP address which the agent wishes to
announce via gARP and for which ARP queries should be answered.

The schema includes fields to support partial reconciliation.

In followup commits, the producer and consumer components will be added.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

dylandreimerink committed:
bpf,maps: Added L2 responder V4 map

Added a new BPF map which is used to communicate to which IPs on which
network interfaces the datapath should respond to ARP requests.

The code agent code to write to this map and the datapath code to read
from this map will be added in a followup commit.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

dylandreimerink committed:
bpf: Added L2 responder datapath logic

This commit adds the logic for the L2 responder. If the feature is
enabled, netdev programs are loaded and attached. When ARP requests
are processed, we check if the combination of network device and IP
address exists in the L2 responder v4 map. If so, send an ARP reply to
the sender with the MAC address of the current network device. Otherwise
pass the ARP request to the stack.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

dylandreimerink committed:
datapath: L2 responder reconciler

This commit adds the pkg/datapath/l2responder. This cell is responsible
for taking the desired state from the L2 announce table and to apply it
to the l2 responder map.

The cell will do partial reconciliation on incremental updates to the
table and periodically perform full reconciliation to catch any drift
from the desired state may it occur for any reason.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

dylandreimerink committed:
Add l2 announcer cell

This commit adds the agent side logic for the L2 announcer feature.
This cell takes all policies, filters out those that do not match the
local node labels. It then takes all services and filters out the
services that do not match the policies. For each remaining service
the agent participates in leader election by way of the k8s leases.

For each service for which we are the leader, we collect all IPs and
network device combinations which should be announced according to the
policies that matched the services in the first place. And this list
gets reconciled to the table with the desired state.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

dylandreimerink committed:
docs: Add documentation for the L2 announcement feature

This commit adds a new page with documentation for users on how to
use the new L2 announcement feature, its requirements, and limitations.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

dylandreimerink committed:
bpf,test: Add the `DUMPCTX` option to the bpf test makefile

The bpf test runner has had a dump context feature for a while now
but there was no makefile option to easily access it. This commit adds
that option which is handy if you wish to inspect the program context
at a binary level during test development or debugging.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

PhilipSchmid commented on Documentation/network/l2-announcements.rst:
```diff
@@ -0,0 +1,314 @@
+************************************
+
+L2 Announcements is a feature which makes services visible and reachable on 
+the local area network. This feature is primarily intended for bare metal 
```
```suggestion
the local area network. This feature is primarily intended for on-premises
```
Isn't it more about on-prem (where it's easy achievable to announce a specific IP via ARP) vs. public cloud, rather than bare metal vs. something? Also, traditional VMs (e.g. hosted on ESXi) are probably even more common than real bare metal nodes.

dylandreimerink replied:
Yes, very true, thx!

PhilipSchmid commented on Documentation/network/l2-announcements.rst:
```diff
@@ -0,0 +1,314 @@
+               --namespace kube-system \\
+               --reuse-values \\
+               --set l2announcements.enabled=true \\
+               --set kubeProxyReplacement=strict
```
This is a non-functional `helm` example. Should we add the following arguments like we do in the [KPR guide](https://docs.cilium.io/en/stable/network/kubernetes/kubeproxy-free/#quick-start)?
```
    --set k8sServiceHost=${API_SERVER_IP} \
    --set k8sServicePort=${API_SERVER_PORT}
```

dylandreimerink replied:
Ah, yes, we should. I copied the Egress gateway example and did not notice this.

PhilipSchmid commented on Documentation/network/l2-announcements.rst:
```diff
@@ -0,0 +1,314 @@
+Prerequisites
+#############
+
+* Kube Proxy replacement mode must be enabled, see :ref:`kubeproxy-free` for details.
```
```suggestion
* Kube Proxy replacement mode must be enabled and set to `strict` mode, see :ref:`kubeproxy-free` for details.
```
I think it would make sense to be precise here. What does "enabled" regarding KPR mean? Is `partial` also fine?

dylandreimerink replied:
I think it also works in some configuration with `partial` mode, but I honestly can't figure out in which combination we do and don't load+attach the `bpf_host` programs which is the main requirement. So I will keep it at strict for now.

PhilipSchmid commented on Documentation/network/l2-announcements.rst:
```diff
@@ -0,0 +1,314 @@
+
+* Kube Proxy replacement mode must be enabled, see :ref:`kubeproxy-free` for details.
+
+* All devices on which L2 Aware LB should be enabled should be included in the 
```
```suggestion
* All devices on which L2 Aware LB should be enabled and included in the 
```

PhilipSchmid commented on Documentation/network/l2-announcements.rst:
```diff
@@ -0,0 +1,314 @@
+Limitations
+###########
+
+* The feature currently does not support IPv6/NDP
```
```suggestion
* The feature currently does not support IPv6/NDP.
```

PhilipSchmid commented on Documentation/network/l2-announcements.rst:
```diff
@@ -0,0 +1,314 @@
+########
+
+Policies provide fine-grained control over which services should be announced,
+where, and how. This is an example of policy using all optional fields:
```
```suggestion
where, and how. This is an example policy using all optional fields:
```

PhilipSchmid commented on Documentation/network/l2-announcements.rst:
```diff
@@ -0,0 +1,314 @@
+which determines which nodes are candidates to announce the services from.
+
+It might be desirable to pick a subset of nodes in you cluster, since the chosen
+node (see :ref:`l2_announcements_leader_election`) will act as the North/South
```
```suggestion
node (see :ref:`l2_announcements_leader_election`) will act as the north/south
```
Let's unify "north/south". Above it was written without capital letters.

PhilipSchmid commented on Documentation/network/l2-announcements.rst:
```diff
@@ -0,0 +1,314 @@
+If ``externalIPs`` is ``true`` all IPs in `.spec.externalIPs <https://kubernetes.io/docs/concepts/services-networking/service/#external-ips>`__
+are announced, which are managed by service authors.
+
+If ``loadBalancerIPs` is ``true`` all IPs in ``.status.loadbalacer.ingress``
```
```suggestion
If ``loadBalancerIPs`` is ``true`` all IPs in ``.status.loadbalacer.ingress``
```

PhilipSchmid commented on Documentation/network/l2-announcements.rst:
```diff
@@ -0,0 +1,314 @@
+------
+
+If a policy is invalid for any number of reasons, the status of the policy will reflect that.
+For example if a invalid match expression is provided:
```
```suggestion
For example if an invalid match expression is provided:
```

PhilipSchmid commented on Documentation/network/l2-announcements.rst:
```diff
@@ -0,0 +1,314 @@
+               --namespace kube-system \\
+               --reuse-values \\
+               --set l2announcements.enabled=true \\
+               --set kubeProxyReplacement=strict \\
```
Same as above. This is basically a non-functional `helm` example. Should we add the following arguments like we do in the [KPR guide](https://docs.cilium.io/en/stable/network/kubernetes/kubeproxy-free/#quick-start)?
```
    --set k8sServiceHost=${API_SERVER_IP} \
    --set k8sServicePort=${API_SERVER_PORT}
```

PhilipSchmid commented on Documentation/network/l2-announcements.rst:
```diff
@@ -0,0 +1,314 @@
+
+* ``l2announcements.leaseRetryPeriod`` if renewing the lease fails, how long 
+  should the agent wait before it tries again. (default is 2s)
+
```
We should perhaps add a note regarding the allowed values ranges. E.g. can `l2announcements.leaseDuration` only be `1s` or more because this value is used for `leaseDurationSeconds`?
Also, what about `leaseRenewDeadline` and `leaseRetryPeriod `: Do both support sub-second configurations (`ms`) or only `leaseRetryPeriod ` (like you show in the `helm upgrade` example below)?

dylandreimerink replied:
I will follow up on that, have to test what actually happens when you go below 1 second, according to the checks in the go-client library its valid: https://github.com/kubernetes/client-go/blob/v0.27.1/tools/leaderelection/leaderelection.go#L77-L97

> Also, what about leaseRenewDeadline and leaseRetryPeriod : Do both support sub-second configurations (ms) or only leaseRetryPeriod  (like you show in the helm upgrade example below)?

They all should support sub second values.

dylandreimerink replied:
Having a duration lower than `leaseDurationSeconds` does not do anything, it just gets rounded to the nearest second. So I will adjust the docs and the code to check for that and to document the limit. Thanks for the sharp eyes.

PhilipSchmid commented on Documentation/network/l2-announcements.rst:
```diff
@@ -0,0 +1,314 @@
+
+When nodes participating in leader election detect that the lease holder did not
+renew the lease for ``leaseDurationSeconds`` amount of seconds, they will ask
+the API server to make them the new holder. The first request to be processed 
```
> When nodes participating in leader election detect that the lease holder did not
renew the lease for ``leaseDurationSeconds`` amount of seconds [...]

Does this detection add some "latency" in addition to `leaseDurationSeconds` for the switch-over process or is this nearly instant because of an API watch operation?

dylandreimerink replied:
If I am reading it right, `renew` calls `tryAquireOrRenew` every `RenewDeadline`: https://github.com/kubernetes/client-go/blob/v0.27.1/tools/leaderelection/leaderelection.go#L262

At that time we check if the lease expired. So in reality the failover is `leaseDurationSeconds` at best and `leaseDurationSeconds + RenewDeadline` at worst.

PhilipSchmid replied:
Alright, thanks for checking. Do you think it's worth mentioning this in the doc?

dylandreimerink replied:
Yea, I think it is, will do!

PhilipSchmid replied:
Awesome, thanks a lot.

PhilipSchmid commented on Documentation/network/l2-announcements.rst:
```diff
@@ -0,0 +1,314 @@
+the API server to make them the new holder. The first request to be processed 
+gets through and the rest are denied.
+
+When a node becomes the leader/holder, it will send out a gratuitous ARP reply 
```
We should probably mention Unsolicited Neighbor Advertisements as well or at least add context that this is IPv4 only. In the limitation section we already mentioned it currently doesn't support IPv6 so perhaps let's go for the latter?

PhilipSchmid commented on pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliuml2announcementpolicies.yaml:
```diff
@@ -0,0 +1,263 @@
+                description: If true, the external IPs of the services are announced
+                type: boolean
+              interfaces:
+                description: A list of regular expressions that express which network
```
```suggestion
                description: A list of regular expressions that express which network interface(s)
```

PhilipSchmid commented on pkg/k8s/apis/cilium.io/v2alpha1/l2announcement_types.go:
```diff
@@ -0,0 +1,97 @@
+	//
+	// +kubebuilder:validation:Optional
+	ExternalIPs bool `json:"externalIPs"`
+	// A list of regular expressions that express which network should be used
```
```suggestion
	// A list of regular expressions that express which network interface(s) should be used
```

PhilipSchmid commented on pkg/k8s/apis/cilium.io/v2alpha1/l2announcement_types.go:
```diff
@@ -0,0 +1,97 @@
+	// +kubebuilder:validation:Optional
+	NodeSelector *slimv1.LabelSelector `json:"nodeSelector"`
+	// ServiceSelector selects a set of services which will be announced over L2 networks
+	//
```
```suggestion
	//
	// If nil this policy applies to all services.
	//
```

PhilipSchmid commented on pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliuml2announcementpolicies.yaml:
```diff
@@ -0,0 +1,263 @@
+                type: object
+              serviceSelector:
+                description: ServiceSelector selects a set of services which will
+                  be announced over L2 networks
```
```suggestion
                  be announced over L2 networks. \n If nil this policy applies to all services.
```

PhilipSchmid reviewed (commented):
Awesome work 🎉! I had a look at the doc, Helm chart, and CRD part and added some minor comments.

raphink commented on Documentation/network/l2-announcements.rst:
```diff
@@ -0,0 +1,323 @@
+load balancing with the service load balancing feature, thus acting as a 
+north/south load balancer.
+
+The advantage over this feature over NodePort services is that each service can
```
```suggestion
The advantage of this feature over NodePort services is that each service can
```

raphink commented on Documentation/network/l2-announcements.rst:
```diff
@@ -0,0 +1,323 @@
+north/south load balancer.
+
+The advantage over this feature over NodePort services is that each service can
+use a unique IP so multiple service can use the same port numbers. When using 
```
```suggestion
use a unique IP so multiple services can use the same port numbers. When using 
```

raphink commented on Documentation/network/l2-announcements.rst:
```diff
@@ -0,0 +1,323 @@
+
+The advantage over this feature over NodePort services is that each service can
+use a unique IP so multiple service can use the same port numbers. When using 
+NodePorts its up to the client to decide to which host to send traffic, if a node
```
```suggestion
NodePorts, it is up to the client to decide to which host to send traffic, and if a node
```

raphink commented on Documentation/network/l2-announcements.rst:
```diff
@@ -0,0 +1,323 @@
+goes down, the IP+Port combo becomes unusable. With L2 announcements the service
+VIP simply migrates to another node and will continue to work.
+
+The L2 Announcements feature and all the requirements can be enabled as follow:
```
```suggestion
The L2 Announcements feature and all the requirements can be enabled as follows:
```

raphink commented on Documentation/network/l2-announcements.rst:
```diff
@@ -0,0 +1,323 @@
+
+* Kube Proxy replacement mode must be enabled and set to `strict` mode, :ref:`kubeproxy-free` for details.
+
+* All devices on which L2 Aware LB should be enabled and included in the 
```
```suggestion
* All devices on which L2 Aware LB will be announced should be enabled and included in the 
```

raphink commented on Documentation/network/l2-announcements.rst:
```diff
@@ -0,0 +1,323 @@
+* All devices on which L2 Aware LB should be enabled and included in the 
+  ``--devices`` flag or ``devices`` Helm option if explicitly set, see :ref:`NodePort Devices`.
+
+* The ``externalIPs.enable=true`` Helm option should be set, if usage of externalIPs
```
```suggestion
* The ``externalIPs.enabled=true`` Helm option should be set, if usage of externalIPs
```

raphink commented on Documentation/network/l2-announcements.rst:
```diff
@@ -0,0 +1,323 @@
+* The feature currently does not support IPv6/NDP.
+
+* Due to the way L3->L2 translation protocols work, one node receives all 
+  per IP, no load balancing before traffic hits the cluster.
```
```suggestion
* Due to the way L3->L2 translation protocols work, one node receives all 
  ARP requests for a specific IP, so no load balancing can happen before traffic hits the cluster.
```

raphink commented on Documentation/network/l2-announcements.rst:
```diff
@@ -0,0 +1,323 @@
+--------
+
+The ``externalIPs`` and ``loadBalancerIPs`` fields determine what sort of IPs 
+are announced. They are ``false`` by default, so a functional policy should always
```
```suggestion
are announced. They are both set to ``false`` by default, so a functional policy should always
```

raphink commented on Documentation/network/l2-announcements.rst:
```diff
@@ -0,0 +1,323 @@
+have one or both set to ``true``.
+
+If ``externalIPs`` is ``true`` all IPs in `.spec.externalIPs <https://kubernetes.io/docs/concepts/services-networking/service/#external-ips>`__
+are announced, which are managed by service authors.
```
```suggestion
If ``externalIPs`` is ``true`` all IPs in the service's `.spec.externalIPs <https://kubernetes.io/docs/concepts/services-networking/service/#external-ips>`__
field are announced. These IPs are are managed by service authors.
```

raphink commented on Documentation/network/l2-announcements.rst:
```diff
@@ -0,0 +1,323 @@
+If ``externalIPs`` is ``true`` all IPs in `.spec.externalIPs <https://kubernetes.io/docs/concepts/services-networking/service/#external-ips>`__
+are announced, which are managed by service authors.
+
+If ``loadBalancerIPs`` is ``true`` all IPs in ``.status.loadbalacer.ingress``
```
```suggestion
If ``loadBalancerIPs`` is ``true`` all IPs in the service's ``.status.loadbalacer.ingress`` field
```

raphink commented on Documentation/network/l2-announcements.rst:
```diff
@@ -0,0 +1,323 @@
+There is a trade-off between fast failure detection and CPU + network usage. 
+Each service incurs a CPU and network overhead, so clusters with smaller amounts
+of services can more easily afford faster failover times. Larger clusters might
+need to increase parameters if the overhead is to much.
```
```suggestion
need to increase parameters if the overhead is too high.
```

raphink commented on Documentation/network/l2-announcements.rst:
```diff
@@ -0,0 +1,323 @@
+gets through and the rest are denied.
+
+When a node becomes the leader/holder, it will send out a gratuitous ARP reply 
+over all of the configured interfaces. Clients whom accept these will update 
```
```suggestion
over all of the configured interfaces. Clients who accept these will update 
```

chancez commented on pkg/datapath/l2responder/l2responder.go:
```diff
@@ -0,0 +1,334 @@
+	}
+
+	for ctx.Err() == nil {
+		maxRev = p.cycle(ctx, maxRev, ticker.C)
```
maxRev here, and above never gets used?

dylandreimerink replied:
maxRev is also an argument. cycle is called in a loop feeding the output from the last call to the next. It is used to track revision numbers for the incremental/partial update

chancez replied:
Ah it's passed into the arguments I see.

qmonnet commented on Documentation/network/l2-announcements.rst:
```diff
@@ -0,0 +1,332 @@
+Prerequisites
+#############
+
+* Kube Proxy replacement mode must be enabled and set to ``strict`` mode, :ref:`kubeproxy-free` for details.
```
```suggestion
* Kube Proxy replacement mode must be enabled and set to ``strict`` mode, see :ref:`kubeproxy-free` for details.
```

qmonnet commented on Documentation/network/l2-announcements.rst:
```diff
@@ -0,0 +1,332 @@
+  ``--devices`` flag or ``devices`` Helm option if explicitly set, see :ref:`NodePort Devices`.
+
+* The ``externalIPs.enabled=true`` Helm option should be set, if usage of externalIPs
+  is desired. Otherwise service load balancing for external IPs is disabled.
```
Terminology: “should” or “must”?

(Same question for the use of “should” at other locations in the doc.)

qmonnet commented on Documentation/network/l2-announcements.rst:
```diff
@@ -0,0 +1,332 @@
+Interfaces
+----------
+
+The interfaces field is a list of regular expressions (`golang syntax <https://pkg.go.dev/regexp/syntax>`__
```
```suggestion
The interfaces field is a list of regular expressions (`golang syntax <https://pkg.go.dev/regexp/syntax>`__)
```

qmonnet commented on Documentation/network/l2-announcements.rst:
```diff
@@ -0,0 +1,332 @@
+expressions will be matched.
+
+L2 announcements only work if the selected devices are also part of the set of 
+devices specified in the ``devices`` Helm option, see :ref:`NodePort Devices`
```
```suggestion
devices specified in the ``devices`` Helm option, see :ref:`NodePort Devices`.
```

qmonnet commented on Documentation/network/l2-announcements.rst:
```diff
@@ -0,0 +1,332 @@
+
+* ``l2announcements.leaseDuration`` determines the ``leaseDurationSeconds`` value
+  of created leases and by extent how long a leader must be "down" before 
+  failover occurs. Its default value is 15s, it must always be greater then 1s
```
`then` -> `than`

qmonnet commented on bpf/bpf_host.c:
```diff
@@ -1228,6 +1275,7 @@ handle_srv6(struct __ctx_buff *ctx)
  * managed by Cilium (e.g., eth0). This program is only attached when:
  * - the host firewall is enabled, or
  * - BPF NodePort is enabled, or
+ * - L2 announcements, or
```
```suggestion
 * - L2 announcements are enabled, or
```

qmonnet commented on bpf/lib/common.h:
```diff
@@ -378,6 +378,11 @@ struct auth_info {
  */
 enum {
 	RUNTIME_CONFIG_UTIME_OFFSET = 0, /* Index to Unix time offset in 512 ns units */
+	RUNTIME_CONFIG_USED_ENTRIES = 1, /* */
```
Missing comment?

qmonnet commented on bpf/tests/pktgen.h:
```diff
@@ -408,6 +422,46 @@ struct ipv6hdr *pktgen__push_default_ipv6hdr(struct pktgen *builder)
+	if (builder->cur_off >= MAX_PACKET_OFF - sizeof(struct arphdreth))
+		return 0;
+
+	layer = ctx_data(ctx) + builder->cur_off;
```
I expect this to fail in the CI now that we're using inline assembly to implement `ctx_data()` and `ctx_data_end()`. If that's the case, please make sure you call these two functions only once (`data = ctx_data(ctx)` then use `data` as much as you want), as the compiler can no longer optimise unnecessary calls out. (You still need to call them again after a helper call that invalidates the call though, but that's not the case here as I read it.)

dylandreimerink replied:
This seems to work, every variable on this line is 64bit so no 32-bit verifier bullshit. Unit tests seem happy in CI as is

qmonnet commented on bpf/tests/tc_l2_announcement.c:
```diff
@@ -0,0 +1,221 @@
+
+#include "common.h"
+
+/* Enable debug output*/
```
```suggestion
/* Enable debug output */
```

qmonnet commented on bpf/tests/tc_l2_announcement.c:
```diff
@@ -0,0 +1,221 @@
+/* Set ETH_HLEN to 14 to indicate that the packet has a 14 byte ethernet header */
+#define ETH_HLEN 14
+
+/* Enable code paths under test*/
```
```suggestion
/* Enable code paths under test */
```

qmonnet commented on bpf/tests/tc_l2_announcement.c:
```diff
@@ -0,0 +1,221 @@
+	test_init();
+
+	data = (void *)(long)ctx->data;
+	data_end = (void *)(long)ctx->data_end;
```
Any reason why not using `ctx_data*()`? Same below.

dylandreimerink replied:
For one the habbit of using `ctx_data` just has not sunken in yet, secondly it was copy-pasta from one of the other tests, likely due to me not using it in the past as well....

qmonnet commented on Documentation/spelling_wordlist.txt:
```diff
@@ -49,6 +49,7 @@ Donenfeld
 Elasticsearch
 FIt
 Facebook
+Failover
```
This one is not needed, the lowcase version below should make the exception case-insensitive.

qmonnet reviewed (commented):
[Looks like I failed to submit my overall comment with the review items above]

So I looked mostly at the documentation, and the BPF code. Looks good overall, with a few nits for the doc and some questions on `ctx_data()` usage in the BPF tests (please see above).

qmonnet reviewed (approved):
Thanks!

borkmann reviewed (approved):
Ack for BPF side, with one small Q: when this enabled together with XDP, the XDP logic will punt this to tc BPF, right?

borkmann reviewed (approved):
Ack for BPF side, with one small Q: when this enabled together with XDP, the XDP logic will punt this to tc BPF, right?

dylandreimerink committed:
CODEOWNERS: Assign new `pkg/l2announcer` package to @cilium/sig-agent

This package needs a codeowner, assinging to @cilium/sig-agent since we
have no specific team for this feature and its a agent component.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

joamaki commented on daemon/cmd/cells.go:
```diff
@@ -61,6 +64,11 @@ var (
 		// Provide option.Config via hive so cells can depend on the agent config.
 		cell.Provide(func() *option.DaemonConfig { return option.Config }),
+
+		statedb.Cell,
```
Let's add a comment describing statedb:

```
// Provides an in-memory transactional database for internal state
```

joamaki commented on daemon/cmd/daemon_main.go:
```diff
@@ -358,6 +359,18 @@ func initializeFlags() {
 	flags.Bool(option.EnableWireguard, false, "Enable wireguard")
 	option.BindEnv(Vp, option.EnableWireguard)
 
+	flags.Bool(option.EnableL2Announcements, false, "Enable L2 announcements")
```
I think it's worth putting these into a config struct in `pkg/l2announcer` instead and then passing that config down to `Daemon` and the `HeaderFileWriter` even if it's a little bit more changes. When would we otherwise get to cleaning these things up?

dylandreimerink replied:
I imagined we would take care of all of these during cleanup of the loader which currently makes fairly heavy use of `AreDevicesRequired`

joamaki replied:
I'm ok with that too.

joamaki commented on pkg/k8s/shared_resources.go:
```diff
@@ -145,3 +147,13 @@ func ciliumCIDRGroup(lc hive.Lifecycle, cs client.Clientset) (resource.Resource[
 	return resource.New[*cilium_api_v2alpha1.CiliumCIDRGroup](lc, lw), nil
 }
+
+func l2AnnouncementPolicyResource(lc hive.Lifecycle, cs client.Clientset) (resource.Resource[*cilium_api_v2alpha1.CiliumL2AnnouncementPolicy], error) {
```
I'd argue that this belongs in the cell of `pkg/l2announcer` as it isn't expected to be used elsewhere.

joamaki commented on pkg/l2announcer/l2announcer.go:
```diff
@@ -0,0 +1,1073 @@
+	"k8s.io/client-go/tools/leaderelection/resourcelock"
+)
+
+var Cell = cell.Module("l2-announcer", "L2 Announcer",
```
nit, I think it's tidier and easier to read if indented this way:
```
var Cell = cell.Module(
  "my-module-id",
  "My module description",

  cell.Provide(New),
)
```

joamaki commented on pkg/l2announcer/l2announcer.go:
```diff
@@ -0,0 +1,1073 @@
+		}
+	}
+
+	for {
```
When do we exit this loop? This is perhaps one downside of writing the tests in terms of the internal methods as we now don't have an easy way to test that we clean up resources on stop with goleak.
Would it be easy enough to add a test that does start+stop and checks no goroutines leak like this?

dylandreimerink replied:
I fixed the not breaking from the loop and I added a test at the end of the _test.go file to check for proper startup/shutdown without leaks.

joamaki commented on pkg/l2announcer/l2announcer.go:
```diff
@@ -0,0 +1,1073 @@
+// This is needed since agents do not track leases for services that we no longer select.
+func (l2a *L2Announcer) leaseGC(ctx context.Context) error {
+	// Can't GC if there is client is disabled
+	if !l2a.Params.Clientset.IsEnabled() {
```
This belongs in the constructor which should then be a no-op if there's no k8s client. Right now we're also missing the check in `run` (it'd crash as the resources would be nil).

joamaki commented on pkg/l2announcer/l2announcer.go:
```diff
@@ -0,0 +1,1073 @@
+// the holder node persists all IPs and netdev combinations selected by the policy to the L2AnnounceTable. Datapath
+// components consume them and handle traffic for the IP+netdev entries.
+type L2Announcer struct {
+	Params l2AnnouncerParams
```
Is there a reason this is exported? Seems a bit suspect to give other packages access to these.

joamaki commented on pkg/l2announcer/l2announcer.go:
```diff
@@ -0,0 +1,1073 @@
+	for _, policy := range l2a.policyStore.List() {
+		err := l2a.upsertPolicy(ctx, policy)
+		if err != nil {
+			errors.Join(errs, fmt.Errorf("make node selector: %w", err))
```
why does this error also have the prefix `make node selector`?

joamaki commented on pkg/l2announcer/l2announcer.go:
```diff
@@ -0,0 +1,1073 @@
+			}
+		}
+		if err := l2a.updatePolicyStatus(ctx, selectedPolicy.policy, "io.cilium/bad-node-selector", nil); err != nil {
+			l2a.Params.Logger.WithError(err).Error("updating policy status failed")
```
Is it safe to continue with the policy deletion when this fails?

dylandreimerink replied:
Yea. This call is meant to correct the status of the policy in case the status indicated the policy was invalid before. If that goes wrong then the status might not be up-to-date, but that should not keep us from changing our desired state.

joamaki commented on pkg/l2announcer/l2announcer.go:
```diff
@@ -0,0 +1,1073 @@
+				if err2 := l2a.updatePolicyStatus(ctx, selectedPolicy.policy, "io.cilium/bad-node-selector", err); err2 != nil {
+					l2a.Params.Logger.WithError(err2).Error("updating policy status failed")
+				}
+				return fmt.Errorf("make node selector: %w", err)
```
`make node selector` seems a bit vague and it's missing the bad `NodeSelector` itself

joamaki commented on pkg/l2announcer/l2announcer.go:
```diff
@@ -0,0 +1,1073 @@
+			return nil
+		})
+		if err != nil {
+			return fmt.Errorf("ProcessEach: %w", err)
```
avoid using function names in errors and be more descriptive, e.g. "failed to recalculate L2 announcer policies: "

joamaki commented on pkg/datapath/tables/cells.go:
```diff
@@ -0,0 +1,12 @@
+	"github.com/cilium/cilium/pkg/hive/cell"
+)
+
+var Cell = cell.Group(
```
I'd make this a module (useful later when graphing just module-level dependencies)

joamaki commented on pkg/datapath/tables/l2_announce.go:
```diff
@@ -0,0 +1,156 @@
+
+func (pne *L2AnnounceEntry) DeepCopy() *L2AnnounceEntry {
+	var n L2AnnounceEntry
+	n.IP = make(net.IP, len(pne.IP))
```
Use `slices.Clone` here too

joamaki commented on pkg/datapath/tables/l2_announce.go:
```diff
@@ -0,0 +1,156 @@
+	Field string
+}
+
+func (rki *resourceKeySliceFieldIndex) FromObject(obj interface{}) (bool, [][]byte, error) {
```
Hm, wonder if we should instead add some generic utilities to deal with types that implement `fmt.Stringer`,
e.g. `StringerFieldIndex[T fmt.Stringer]` and `StringerSliceFieldIndex[TS []T, T fmt.Stringer]`. Anyway, more food for thought for now.

joamaki commented on pkg/l2announcer/l2announcer.go:
```diff
@@ -0,0 +1,1073 @@
+	for {
+		select {
+		case <-ctx.Done():
+		case event := <-svcChan:
```
We don't process the channel closing. I think it's fine as long as we bail out from `ctx.Done`. Since we don't currently, I expect this turns into a busy-loop when one of the channels close.

joamaki commented on pkg/statedb/index.go:
```diff
@@ -0,0 +1,152 @@
+	}
+	stringer, ok := args[0].(fmt.Stringer)
+	if !ok {
+		return nil, fmt.Errorf("argument must be a string: %#v", args[0])
```
s/string/fmt.Stringer/

joamaki commented on pkg/datapath/l2responder/l2responder.go:
```diff
@@ -0,0 +1,334 @@
+	"github.com/vishvananda/netlink"
+)
+
+var Cell = cell.Module("l2-responder", "L2 Responder Reconciler",
```
Could you add a comment to explain what this does in some detail so a reader can understand this.

Perhaps improve the module description to describe what it does? Current description is too jargon heavy. Also no need to repeat "L2 Responder" in description when it's in the module id already.

joamaki commented on pkg/datapath/l2responder/l2responder.go:
```diff
@@ -0,0 +1,334 @@
+	arMap := p.params.L2ResponderMap
+	lr := cachingLinkResolver{nl: p.params.NetLink}
+
+	log.Info("l2 announcer table full reconciliation")
```
Is this really a useful thing to log at Info? Seems more like Debug to me. Same in partial reconciliation. 

If this would be at info level then I'd log at the end and include some stats about what was done. This by itself seems a bit useless.

joamaki commented on pkg/datapath/tables/cells.go:
```diff
@@ -0,0 +1,14 @@
+
+var Cell = cell.Module(
+	"datapath-tables",
+	"Datapath tables",
```
Don't repeat module id in description, but rather describe what the module does or provides. No need to repeat module id (they'll always be shown next to each other). E.g. now this shows up as something silly like:

```
datapath-tables: Datapath tables
```

joamaki commented on pkg/datapath/tables/l2_announce.go:
```diff
@@ -0,0 +1,91 @@
+	Revision uint64
+}
+
+func (pne *L2AnnounceEntry) DeepCopy() *L2AnnounceEntry {
```
This does not copy `Deleted` nor `Revision`. How about `copy := *pne` and then clone the slices?

joamaki commented on pkg/datapath/agentliveness/agent_liveness.go:
```diff
@@ -0,0 +1,67 @@
+	"github.com/spf13/pflag"
+)
+
+var Cell = cell.Group(
```
Use module instead here. Rational is that I want module-level dependency graphs to be useful and this is complex enough with multiple dependencies that I want this to show up as it's own thing.

joamaki commented on pkg/datapath/cells.go:
```diff
@@ -48,6 +51,14 @@ var Cell = cell.Module(
+	// the datapath of the liveness of the agent.
+	agentliveness.Cell,
+
+	l2responder.Cell,
```
One-line comment here what this does. Same for tables.

joamaki commented on pkg/l2announcer/l2announcer.go:
```diff
@@ -0,0 +1,1113 @@
+
+var Cell = cell.Module(
+	"l2-announcer",
+	"L2 Announcer",
```
Don't repeat module id. Do a proper description of what this does. And add detailed comment for `Cell` to describe the functionality in more detail.

joamaki commented on pkg/l2announcer/l2announcer.go:
```diff
@@ -0,0 +1,1113 @@
+	select {
+	// Send the change to the central routine to avoid race conditions
+	case l2a.devicesChan <- devices:
+	default:
```
We'll end up dropping notifications here if we exhaust the buffer. That's pretty scary. We can't block either with the current structure where `run` isn't spawned if there's no k8s client.

I'd propose to make this blocking send, but in the case of `!params.Clientset.IsEnabled` we return a no-op implementation from `NewL2Announcer`. Later with the devices table we can get rid of the provider completely for `L2Announcer` as we won't have a public API.

dylandreimerink replied:
Fair point. We only really care about the latest update anyway. So now the function will update the slice directly which can technically cause a race-condition. But I also send a "trigger" on the channel which should take care of any issues that might have resulted from modifying the devices list halfway through. This approach is nice since it will coalesce changes if multiple happen in short succession.

joamaki reviewed (approved):
LGTM, left some nits :)

bimmlerd reviewed (approved):
contributing changes lgtm

michi-covalent reviewed (approved):
lgtm for cli files.

joestringer commented on pkg/datapath/l2responder/l2responder.go:
```diff
@@ -0,0 +1,340 @@
+func (p *l2ResponderReconciler) run(ctx context.Context) error {
+	log := p.params.Logger
+
+	// This timer triggers full reconciliation once in a while, in case partial reconciliation
+	// got out of sync or the map was changed underneath us.
+	ticker := time.NewTicker(5 * time.Minute)
```
@dylandreimerink Did you consider using something more formal for reconciliation? We have facilities like `pkg/controller` or Hive jobs in order to support periodic reconcilers like this while also exposing the status into `cilium status` output in a standardized way so that they're easer to debug.

cc @joamaki

dylandreimerink replied:
We have. We are playing with a few ideas to find the best solution: https://github.com/cilium/cilium/pull/27884 and https://github.com/cilium/cilium/pull/28276

Not just for L2 but a bit broader
